File: shell-ios/CloudGuard/JsChannel.swift
import Foundation
import WebKit

/// JS 通道：封装「原生 -> JS」的两类回传（对应 Android 端 JsChannel）。
///
///  1) 调用回传： window.__nativeCallback(callbackId, { code, data, msg })
///     - code === 0 成功，其余失败
///  2) 事件推送： window.__nativeEmit(eventName, data)
///
/// 所有回传统一在主线程通过 evaluateJavaScript 执行。
/// 注意：第二个参数为「JSON 字符串」（与 utils/h5-native-bridge.js 的解析约定一致）。
final class JsChannel {

    private weak var webView: WKWebView?

    init(webView: WKWebView?) {
        self.webView = webView
    }

    func attach(_ webView: WKWebView) {
        self.webView = webView
    }

    func resolve(_ callbackId: String?, code: Int, data: [String: Any]? = nil, msg: String? = nil) {
        guard let cb = callbackId, !cb.isEmpty else { return }
        var res: [String: Any] = ["code": code]
        if let msg = msg { res["msg"] = msg }
        if let data = data { res["data"] = data }
        let json = JsChannel.encode(res)
        evaluate("window.__nativeCallback(\(JsChannel.stringLiteral(cb)), \(JsChannel.stringLiteral(json)));")
    }

    func resolveOk(_ callbackId: String?, _ data: [String: Any]? = nil) {
        resolve(callbackId, code: 0, data: data)
    }

    func resolveErr(_ callbackId: String?, _ msg: String, code: Int = -1) {
        resolve(callbackId, code: code, data: nil, msg: msg)
    }

    func emit(_ event: String, _ data: [String: Any]) {
        let json = JsChannel.encode(data)
        evaluate("window.__nativeEmit(\(JsChannel.stringLiteral(event)), \(JsChannel.stringLiteral(json)));")
    }

    private func evaluate(_ js: String) {
        DispatchQueue.main.async { [weak self] in
            guard let webView = self?.webView else { return }
            webView.evaluateJavaScript(js, completionHandler: nil)
        }
    }

    /// 对象 -> JSON 字符串
    static func encode(_ obj: Any) -> String {
        if let data = try? JSONSerialization.data(withJSONObject: obj, options: []),
           let s = String(data: data, encoding: .utf8) {
            return s
        }
        return "{}"
    }

    /// 字符串 -> JS 字符串字面量（含引号）
    static func stringLiteral(_ s: String) -> String {
        if let data = try? JSONSerialization.data(withJSONObject: [s], options: []),
           var str = String(data: data, encoding: .utf8) {
            str.removeFirst()
            str.removeLast()
            return str
        }
        return "\"\""
    }
}

File: shell-android/app/src/main/java/com/yefiot/cloudguard/bridge/JsChannel.kt
package com.yefiot.cloudguard.bridge

import android.os.Handler
import android.os.Looper
import android.webkit.WebView
import org.json.JSONObject

/**
 * JS 通道：封装「原生 -> JS」的两类回传。
 *
 *  1) 调用回传： window.__nativeCallback(callbackId, { code, data, msg })
 *     - code === 0 成功，其余失败
 *  2) 事件推送： window.__nativeEmit(eventName, data)
 *
 * 所有回传统一在主线程通过 evaluateJavascript 执行，避免 WebView 线程问题。
 */
class JsChannel(private val webView: WebView) {

    private val handler = Handler(Looper.getMainLooper())

    fun resolve(callbackId: String?, code: Int, data: JSONObject? = null, msg: String? = null) {
        if (callbackId.isNullOrEmpty()) return
        val res = JSONObject()
        res.put("code", code)
        if (msg != null) res.put("msg", msg)
        if (data != null) res.put("data", data)
        evaluate("window.__nativeCallback(${JSONObject.quote(callbackId)}, ${JSONObject.quote(res.toString())});")
    }

    fun resolveOk(callbackId: String?, data: JSONObject? = null) = resolve(callbackId, 0, data)

    fun resolveErr(callbackId: String?, msg: String, code: Int = -1) = resolve(callbackId, code, null, msg)

    fun emit(event: String, data: JSONObject) {
        evaluate("window.__nativeEmit(${JSONObject.quote(event)}, ${JSONObject.quote(data.toString())});")
    }

    private fun evaluate(js: String) {
        handler.post {
            try {
                webView.evaluateJavascript(js, null)
            } catch (_: Throwable) {
                // WebView 已销毁等情况静默忽略
            }
        }
    }
}
